File: bank-service/src/main/java/rs/edu/raf/controller/CardController.java
package rs.edu.raf.controller;

import io.swagger.annotations.ApiOperation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import rs.edu.raf.dto.CardDTO;
import rs.edu.raf.dto.CreateCardDto;
import rs.edu.raf.service.CardService;

import java.util.List;


@RestController
@RequestMapping("/cards")
@AllArgsConstructor
@Tag(name = "Cards", description = "Managing cards")
@SecurityRequirement(name = "jwt")
@CrossOrigin(origins = "*")
public class CardController {
    private final CardService cardService;

    @ApiOperation(value = "Create card")
    @PostMapping("/create")
    public ResponseEntity<CardDTO> createCard(@RequestBody @Valid CreateCardDto createCardDto){
        return new ResponseEntity<>(cardService.createCard(createCardDto), HttpStatus.CREATED);
    }

    @ApiOperation(value = "Block card")
    @PatchMapping("/block/{cardNumber}")
    public ResponseEntity<?> blockCard(@Parameter(name = "Card number") @PathVariable("cardNumber") String cardNumber){
        cardService.blockCard(cardNumber);
        return new ResponseEntity<>(HttpStatus.OK);
    }

    @ApiOperation(value = "Unblock card")
    @PatchMapping("/unblock/{cardNumber}")
    public ResponseEntity<?> unblockCard(@Parameter(name = "Card number") @PathVariable("cardNumber") String cardNumber){
        cardService.unblockCard(cardNumber);
        return new ResponseEntity<>(HttpStatus.OK);
    }

    @ApiOperation(value = "Deactivate card")
    @PatchMapping("/deactivate/{cardNumber}")
    public ResponseEntity<?> deactivateCard(@Parameter(name = "Card number") @PathVariable("cardNumber") String cardNumber){
        cardService.deactivateCard(cardNumber);
        return new ResponseEntity<>(HttpStatus.OK);
    }


    @ApiOperation(value = "Get all cards")
    @GetMapping
    public ResponseEntity<List<CardDTO>> getAllCards(){
        return new ResponseEntity<>(cardService.getAllCards(), HttpStatus.OK);
    }

    @ApiOperation(value = "Get all cards for specified client")
    @GetMapping("/client/{id}")
    public ResponseEntity<List<CardDTO>> getAllCardsForClient(@Parameter(name = "Client ID") @PathVariable("id") Long id){
        return new ResponseEntity<>(cardService.getAllCardsForUser(id), HttpStatus.OK);
    }

    @ApiOperation(value = "Get cards for currently logged client")
    @GetMapping("/client")
    public ResponseEntity<List<CardDTO>> getAllCardsForLoggedClient(@RequestAttribute("userId") Long id){
        return new ResponseEntity<>(cardService.getAllCardsForUser(id), HttpStatus.OK);
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/controller/CardController.kt
package org.example.bankservicekotlin.controller

import jakarta.validation.Valid
import org.example.bankservicekotlin.dto.CardDTO
import org.example.bankservicekotlin.dto.CreateCardDTO
import org.example.bankservicekotlin.security.JwtUtil
import org.example.bankservicekotlin.service.CardService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/kotlin/cards")
class CardController(
    private val cardService: CardService,
    private val jwtUtil: JwtUtil
) {
    @PostMapping("/create")
    fun createCard(@RequestBody createCardDto: @Valid CreateCardDTO): ResponseEntity<CardDTO> {
        return ResponseEntity<CardDTO>(cardService.createCard(createCardDto), HttpStatus.CREATED)
    }

    @PatchMapping("/block/{cardNumber}")
    fun blockCard(@PathVariable("cardNumber") cardNumber: String): ResponseEntity<*> {
        cardService.blockCard(cardNumber)
        return ResponseEntity<Any>(HttpStatus.OK)
    }

    @PatchMapping("/unblock/{cardNumber}")
    fun unblockCard(@PathVariable("cardNumber") cardNumber: String): ResponseEntity<*> {
        cardService.unblockCard(cardNumber)
        return ResponseEntity<Any>(HttpStatus.OK)
    }

    @PatchMapping("/deactivate/{cardNumber}")
    fun deactivateCard(@PathVariable("cardNumber") cardNumber: String): ResponseEntity<*> {
        cardService.deactivateCard(cardNumber)
        return ResponseEntity<Any>(HttpStatus.OK)
    }

    @GetMapping
    fun getAllCards(): ResponseEntity<List<CardDTO>> {
        return ResponseEntity<List<CardDTO>>(cardService.getAllCards(), HttpStatus.OK)
    }

    @GetMapping("/client/{id}")
    fun getAllCardsForClient(@PathVariable("id") id: Long): ResponseEntity<List<CardDTO>> {
        return ResponseEntity<List<CardDTO>>(cardService.getAllCardsForUser(id), HttpStatus.OK)
    }

    @GetMapping("/client")
    fun getAllCardsForLoggedClient(): ResponseEntity<List<CardDTO>> {
        return ResponseEntity<List<CardDTO>>(cardService.getAllCardsForUser(jwtUtil.getIDForLoggedUser()), HttpStatus.OK)
    }
}